### ISSUE TITLE
[Core] Error on Windows with virtual environment: gRPC unavailable

### ISSUE THREAD

**Issue by Jet132** [ray-project/ray#23945]
### What happened + What you expected to happen

When running Ray locally on Windows 10:

```python
ray.init()

# Some distributed task
tune.run(...)
```

it runs for 30 seconds and then terminates with the following error:

```
2022-04-15 18:40:10,052	ERROR ray_trial_executor.py:102 -- An exception occurred when trying to stop the Ray actor:Traceback (most recent call last):
  File "<python>\lib\site-packages\ray\tune\ray_trial_executor.py", line 93, in post_stop_cleanup
    ray.get(future, timeout=0)
  File "<python>\lib\site-packages\ray\_private\client_mode_hook.py", line 105, in wrapper
    return func(*args, **kwargs)
  File "<python>\lib\site-packages\ray\worker.py", line 1811, in get
    raise value
ray.exceptions.RayActorError: The actor died unexpectedly before finishing this task.
	class_name: ImplicitFunc
	actor_id: c422fd8fc896ec3e3073fb2201000000
	pid: 17256
	namespace: 88cbc294-b174-407a-98b8-4d963422ff38
	ip: 127.0.0.1
The actor is dead because its owner has died. Owner Id: 01000000ffffffffffffffffffffffffffffffffffffffffffffffff Owner Ip address: 127.0.0.1 Owner worker exit type: NODE_DIED

Traceback (most recent call last):
  File "<redacted program>", line 125, in <module>
    asyncio.run(main())
  File "c:\python39\lib\asyncio\runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "c:\python39\lib\asyncio\base_events.py", line 642, in run_until_complete
    return future.result()
  File "<redacted program>", line 75, in main
    analysis = tune.run(
  File "<python>\lib\site-packages\ray\tune\tune.py", line 695, in run
    raise TuneError("Trials did not complete", incomplete_trials)
ray.tune.error.TuneError: ('Trials did not complete', [_tune_train_9582e_00000])
```

Viewing the Ray logs the gRPC unavailable error comes up multiple times in different log files. The most detailed one can be found in `dashboard.log`:

```
2022-04-15 18:40:04,364	ERROR node_head.py:259 -- Error updating node stats of 5241012cc8ceeec135c64c837765766475f2416a95c21496eabd871c.
Traceback (most recent call last):
  File "<python>\lib\site-packages\ray\dashboard\modules\node\node_head.py", line 250, in _update_node_stats
    reply = await stub.GetNodeStats(
  File "<python>\lib\site-packages\grpc\aio\_call.py", line 290, in __await__
    raise _create_rpc_error(self._cython_call._initial_metadata,
grpc.aio._call.AioRpcError: <AioRpcError of RPC that terminated with:
	status = StatusCode.UNAVAILABLE
	details = "failed to connect to all addresses"
	debug_error_string = "{"created":"@1650040804.364000000","description":"Failed to pick subchannel","file":"src/core/ext/filters/client_channel/client_channel.cc","file_line":3135,"referenced_errors":[{"created":"@1650040804.364000000","description":"failed to connect to all addresses","file":"src/core/lib/transport/error_utils.cc","file_line":163,"grpc_status":14}]}"
>
```

Even with the Windows firewall disabled it doesn't work. Though it does work when running it in WSL Ubuntu-18.04.

### Versions / Dependencies

Relevant Python `3.9.7` libraries explicitly installed via Pipenv `2020.11.15`:
```
ray[default, tune]==1.12.0
```
Windows 10 Pro version: `19044.1645`

### Reproduction script

In my case, I used a Pipenv virtual environment but I got the same result when testing with Venv:

```shell
pip install --user pipenv
# Need to manually install redis for ray to start
pipenv install ray[default] redis
pipenv run python ./main.py
```
Contents of `main.py`:

```python
import time
import ray

ray.init()
time.sleep(30)
```

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

**Comment by czgdp1807** [comment]
The script in the OP works fine with the `master` (f72555262afbbfc1aabb87c9e40839aaaee3ba0b) branch. Below are the logs,

```
2022-04-30 08:03:33,375 INFO services.py:1470 -- View the Ray dashboard at http://127.0.0.1:8265
2022-04-30 08:03:42,610 WARNING function_runner.py:603 -- Function checkpointing is disabled. This may result in unexpected behavior when using checkpointing features or certain schedulers. To enable, set the train function arguments to be `func(config, checkpoint_dir=None)`.
2022-04-30 08:03:46,032 ERROR syncer.py:147 -- Log sync requires rsync to be installed.
== Status ==
Current time: 2022-04-30 08:03:56 (running for 00:00:10.69)
Memory usage on this node: 7.6/16.0 GiB
Using HyperBand: num_stopped=0 total_brackets=1
Round #0:
  Bracket(Max Size (n)=81, Milestone (r)=1, completed=0.0%): {RUNNING: 1}
Resources requested: 1.0/8 CPUs, 0/0 GPUs, 0.0/5.53 GiB heap, 0.0/2.76 GiB objects
Result logdir: C:\Users\gagan\ray_results\run_trail_2022-04-30_08-03-42
Number of trials: 1/1 (1 RUNNING)
+-----------------------+----------+----------------+----------+
| Trial name            | status   | loc            |   metric |
|-----------------------+----------+----------------+----------|
| run_trail_0d3b5_00000 | RUNNING  | 127.0.0.1:8944 | 0.199042 |
+-----------------------+----------+----------------+----------+


== Status ==
Current time: 2022-04-30 08:04:01 (running for 00:00:15.75)
Memory usage on this node: 7.6/16.0 GiB
Using HyperBand: num_stopped=0 total_brackets=1
Round #0:
  Bracket(Max Size (n)=81, Milestone (r)=1, completed=0.0%): {RUNNING: 1}
Resources requested: 1.0/8 CPUs, 0/0 GPUs, 0.0/5.53 GiB heap, 0.0/2.76 GiB objects
Result logdir: C:\Users\gagan\ray_results\run_trail_2022-04-30_08-03-42
Number of trials: 1/1 (1 RUNNING)
+-----------------------+----------+----------------+----------+
| Trial name            | status   | loc            |   metric |
|-----------------------+----------+----------------+----------|
| run_trail_0d3b5_00000 | RUNNING  | 127.0.0.1:8944 | 0.199042 |
+-----------------------+----------+----------------+----------+


Result for run_trail_0d3b5_00000:
  date: 2022-04-30_08-04-06
  done: false
  experiment_id: 871537a28ca941fba7ab5f37486cd920
  hostname: quansight-anysc
  iterations_since_restore: 1
  metric: 0.19904181573774615
  node_ip: 127.0.0.1
  pid: 8944
  time_since_restore: 10.01623272895813
  time_this_iter_s: 10.01623272895813
  time_total_s: 10.01623272895813
  timestamp: 1651305846
  timesteps_since_restore: 0
  training_iteration: 1
  trial_id: 0d3b5_00000
  warmup_time: 0.03125166893005371

== Status ==
Current time: 2022-04-30 08:04:06 (running for 00:00:20.88)
Memory usage on this node: 7.6/16.0 GiB
Using HyperBand: num_stopped=0 total_brackets=1
Round #0:
  Bracket(Max Size (n)=81, Milestone (r)=1, completed=0.2%): {PAUSED: 1}
Resources requested: 0/8 CPUs, 0/0 GPUs, 0.0/5.53 GiB heap, 0.0/2.76 GiB objects
Result logdir: C:\Users\gagan\ray_results\run_trail_2022-04-30_08-03-42
Number of trials: 1/1 (1 PAUSED)
+-----------------------+----------+----------------+----------+--------+------------------+----------+
| Trial name            | status   | loc            |   metric |   iter |   total time (s) |   metric |
|-----------------------+----------+----------------+----------+--------+------------------+----------|
| run_trail_0d3b5_00000 | PAUSED   | 127.0.0.1:8944 | 0.199042 |      1 |          10.0162 | 0.199042 |
+-----------------------+----------+----------------+----------+--------+------------------+----------+


(pid=) 2022-04-30 08:04:06,282  INFO context.py:67 -- Exec'ing worker with command: "C:\ProgramData\Anaconda3\envs\ray_dev\python.exe" c:\users\gagan\ray_project\ray\python\ray\workers/default_worker.py --node-ip-address=127.0.0.1 --node-manager-port=52506 --object-store-name=tcp://127.0.0.1:52650 --raylet-name=tcp://127.0.0.1:61915 --redis-address=None --storage=None --temp-dir=C:\CONDA_~1\ray --metrics-agent-port=54796 --logging-rotate-bytes=536870912 --logging-rotate-backup-count=5 --gcs-address=127.0.0.1:59717 --redis-password={removed} --startup-token=8 --runtime-env-hash=-1595412440
== Status ==
Current time: 2022-04-30 08:04:11 (running for 00:00:26.00)
Memory usage on this node: 7.5/16.0 GiB
Using HyperBand: num_stopped=0 total_brackets=1
Round #0:
  Bracket(Max Size (n)=27, Milestone (r)=3, completed=0.2%): {RUNNING: 1}
Resources requested: 1.0/8 CPUs, 0/0 GPUs, 0.0/5.53 GiB heap, 0.0/2.76 GiB objects
Result logdir: C:\Users\gagan\ray_results\run_trail_2022-04-30_08-03-42
Number of trials: 1/1 (1 RUNNING)
+-----------------------+----------+----------------+----------+--------+------------------+----------+
| Trial name            | status   | loc            |   metric |   iter |   total time (s) |   metric |
|-----------------------+----------+----------------+----------+--------+------------------+----------|
| run_trail_0d3b5_00000 | RUNNING  | 127.0.0.1:8944 | 0.199042 |      1 |          10.0162 | 0.199042 |
+-----------------------+----------+----------------+----------+--------+------------------+----------+


(run_trail pid=8912) 2022-04-30 08:04:16,361    INFO trainable.py:588 -- Restored on 127.0.0.1 from checkpoint: C:\Users\gagan\ray_results\run_trail_2022-04-30_08-03-42\run_trail_0d3b5_00000_0_metric=0.1990_2022-04-30_08-03-45\checkpoint_tmpfe16b7\.\
(run_trail pid=8912) 2022-04-30 08:04:16,377    INFO trainable.py:597 -- Current state after restoring: {'_iteration': 0, '_timesteps_total': 0, '_time_total': 10.01623272895813, '_episodes_total': 0}
== Status ==
Current time: 2022-04-30 08:04:16 (running for 00:00:31.11)
Memory usage on this node: 7.6/16.0 GiB
Using HyperBand: num_stopped=0 total_brackets=1
Round #0:
  Bracket(Max Size (n)=27, Milestone (r)=3, completed=0.2%): {RUNNING: 1}
Resources requested: 1.0/8 CPUs, 0/0 GPUs, 0.0/5.53 GiB heap, 0.0/2.76 GiB objects
Result logdir: C:\Users\gagan\ray_results\run_trail_2022-04-30_08-03-42
Number of trials: 1/1 (1 RUNNING)
+-----------------------+----------+----------------+----------+--------+------------------+----------+
| Trial name            | status   | loc            |   metric |   iter |   total time (s) |   metric |
|-----------------------+----------+----------------+----------+--------+------------------+----------|
| run_trail_0d3b5_00000 | RUNNING  | 127.0.0.1:8944 | 0.199042 |      1 |          10.0162 | 0.199042 |
+-----------------------+----------+----------------+----------+--------+------------------+----------+


== Status ==
Current time: 2022-04-30 08:04:21 (running for 00:00:36.16)
Memory usage on this node: 7.6/16.0 GiB
Using HyperBand: num_stopped=0 total_brackets=1
Round #0:
  Bracket(Max Size (n)=27, Milestone (r)=3, completed=0.2%): {RUNNING: 1}
Resources requested: 1.0/8 CPUs, 0/0 GPUs, 0.0/5.53 GiB heap, 0.0/2.76 GiB objects
Result logdir: C:\Users\gagan\ray_results\run_trail_2022-04-30_08-03-42
Number of trials: 1/1 (1 RUNNING)
+-----------------------+----------+----------------+----------+--------+------------------+----------+
| Trial name            | status   | loc            |   metric |   iter |   total time (s) |   metric |
|-----------------------+----------+----------------+----------+--------+------------------+----------|
| run_trail_0d3b5_00000 | RUNNING  | 127.0.0.1:8944 | 0.199042 |      1 |          10.0162 | 0.199042 |
+-----------------------+----------+----------------+----------+--------+------------------+----------+


2022-04-30 08:04:26,393 INFO hyperband.py:465 -- Restoring from a previous point in time. Previous=1; Now=1
Result for run_trail_0d3b5_00000:
  date: 2022-04-30_08-04-26
  done: false
  episodes_total: 0
  experiment_id: 871537a28ca941fba7ab5f37486cd920
  hostname: quansight-anysc
  iterations_since_restore: 1
  metric: 0.19904181573774615
  node_ip: 127.0.0.1
  pid: 8912
  time_since_restore: 10.016031742095947
  time_this_iter_s: 10.016031742095947
  time_total_s: 20.032264471054077
  timestamp: 1651305866
  timesteps_since_restore: 0
  timesteps_total: 0
  training_iteration: 1
  trial_id: 0d3b5_00000
  warmup_time: 0.03127408027648926

Result for run_trail_0d3b5_00000:
  date: 2022-04-30_08-04-26
  done: true
  episodes_total: 0
  experiment_id: 871537a28ca941fba7ab5f37486cd920
  experiment_tag: 0_metric=0.1990
  hostname: quansight-anysc
  iterations_since_restore: 1
  metric: 0.19904181573774615
  node_ip: 127.0.0.1
  pid: 8912
  time_since_restore: 10.016031742095947
  time_this_iter_s: 10.016031742095947
  time_total_s: 20.032264471054077
  timestamp: 1651305866
  timesteps_since_restore: 0
  timesteps_total: 0
  training_iteration: 1
  trial_id: 0d3b5_00000
  warmup_time: 0.03127408027648926

== Status ==
Current time: 2022-04-30 08:04:26 (running for 00:00:41.14)
Memory usage on this node: 7.6/16.0 GiB
Using HyperBand: num_stopped=0 total_brackets=1
Round #0:
  Bracket(Max Size (n)=9, Milestone (r)=9, completed=0.2%): {TERMINATED: 1}
Resources requested: 0/8 CPUs, 0/0 GPUs, 0.0/5.53 GiB heap, 0.0/2.76 GiB objects
Result logdir: C:\Users\gagan\ray_results\run_trail_2022-04-30_08-03-42
Number of trials: 1/1 (1 TERMINATED)
+-----------------------+------------+----------------+----------+--------+------------------+------+----------+
| Trial name            | status     | loc            |   metric |   iter |   total time (s) |   ts |   metric |
|-----------------------+------------+----------------+----------+--------+------------------+------+----------|
| run_trail_0d3b5_00000 | TERMINATED | 127.0.0.1:8912 | 0.199042 |      1 |          20.0323 |    0 | 0.199042 |
+-----------------------+------------+----------------+----------+--------+------------------+------+----------+


2022-04-30 08:04:26,643 INFO tune.py:747 -- Total run time: 44.06 seconds (41.11 seconds for the tuning loop).
(pid=) 2022-04-30 08:04:26,549  INFO context.py:67 -- Exec'ing worker with command: "C:\ProgramData\Anaconda3\envs\ray_dev\python.exe" c:\users\gagan\ray_project\ray\python\ray\workers/default_worker.py --node-ip-address=127.0.0.1 --node-manager-port=52506 --object-store-name=tcp://127.0.0.1:52650 --raylet-name=tcp://127.0.0.1:61915 --redis-address=None --storage=None --temp-dir=C:\CONDA_~1\ray --metrics-agent-port=54796 --logging-rotate-bytes=536870912 --logging-rotate-backup-count=5 --gcs-address=127.0.0.1:59717 --redis-password={removed} --startup-token=9 --runtime-env-hash=-1595412440
```

I think this is connected with https://github.com/ray-project/ray/pull/23223. Not fully sure though. After merging with `master` workers get enough time to register themselves in the worker pool so may be that's why the script in the OP worked with the `master` branch build.

**Comment by Jet132** [comment]
Can confirm it works. Installed today's nightly and it's successfully training. Thank you very much 👍

**Comment by Jet132** [comment]
Nvm I was too quick to judge. Still the same error, even with the 2 newest passing commits (627b9f2e888b05434bb67f547b390409f26538e7 and f72555262afbbfc1aabb87c9e40839aaaee3ba0b) on `master`.

**Comment by czgdp1807** [comment]
I executed the script multiple times on my system and it worked fine all the time. My Azure Windows has access to 8 CPUs and 16 GB RAM.

**Comment by Jet132** [comment]
I've run the script also multiple times on two separate machines (same windows versions, 32GB RAM, min 8 CPUs, 1 GPU) but to no avail. 

With what systems does gRPC interact? Might be useful information as I could give a more detailed environment description and check for misconfigurations.

**Comment by mattip** [comment]
The errors are being logged in the gRPC component, but they indicate that workers cannot be contacted. There are many causes for this. One is that it takes too much time to start the worker, so the monitoring process times out and reports an error back to the HEAD process. This is usually what I see when there is a `failed to connect to all addresses` message. Especially suspicious is the report that the first line of output you reported is 
```
2022-04-15 18:40:10,052	ERROR ray_trial_executor.py:102 -- \
    An exception occurred when trying to stop the Ray actor:Traceback (most recent call last):
```

The whole task should take less than the 30 seconds you report for just that line to appear.

I note you are using pipenv
> Relevant Python 3.9.7 libraries explicitly installed via pipenv 2020.11.15:

How exactly are you setting this up, starting from opening a DOS box window? Maybe something in the interaction between pipenv and the environment variables it sets (`PATH`, `PYTHONPATH`, etc) are causing worker startup to fail on windows?

**Comment by Jet132** [comment]
Retested it with venv and an increased sleep of 30 seconds (updated the OP) with the same result. I'm using a normal pipenv installation (`pip install pipenv --user`) and run the program inside the `pipenv shell`. One thing to note is that I have to manually install the `redis` package to even be able to start the program.

Here are the notable environment variables on the main and worker process:
```env
PATH=C:\Users\<user>\.virtualenvs\Test-15efavF3/Scripts;C:\Python39\Scripts\;C:\Python39\;C:\Python310\Scripts\;C:\Python310\;<more programs>
RAY_CLIENT_MODE=0
TEMP=C:\Users\<user>\AppData\Local\Temp
TMP=C:\Users\<user>\AppData\Local\Temp
VIRTUAL_ENV=C:\Users\<user>\.virtualenvs\Test-15efavF3
```
The worker process has some additional variables:
```env
PROMPT=$P$G
PYTHONBREAKPOINT=ray.util.rpdb.set_trace
RAY_JOB_ID=01000000
RAY_RAYLET_PID=20352
SPT_NOENV=1
TUNE_ORIG_WORKING_DIR=C:\Users\<user>\Downloads\Test
```

Judging by the `python-core-worker-*.log` logs all the workers start up to 4 seconds after the driver. The actor itself gets created on the worker about 5 seconds after the first driver log entry. Here are the relevant log files:
- [python-core-driver-01000000ffffffffffffffffffffffffffffffffffffffffffffffff_19808.log](https://github.com/ray-project/ray/files/8613004/python-core-driver-01000000ffffffffffffffffffffffffffffffffffffffffffffffff_19808.log)
- [python-core-worker-09a2c045086ceb3c8a99d43cc16b0fd78ec3eabc3598edfbc1d05e0c_1888.log](https://github.com/ray-project/ray/files/8613011/python-core-worker-09a2c045086ceb3c8a99d43cc16b0fd78ec3eabc3598edfbc1d05e0c_1888.log)
- [raylet.out](https://github.com/ray-project/ray/files/8613034/raylet.out.txt)
- [dashboard_agent.log](https://github.com/ray-project/ray/files/8613052/dashboard_agent.log)

**Comment by mattip** [comment]
> `PATH=C:\Users\<user>\.virtualenvs\Test-15efavF3/Scripts;C:\Python39\Scripts\;C:\Python39\;C:\Python310\Scripts\;C:\Python310\;<more programs>`

Does anything change if you remove `Python310` from your path?

> I'm using a normal pipenv installation (`pip install pipenv --user`) and run the program inside the pipenv shell
Sorry, I am not a pipenv user. Can you explain the steps to install ray?
```
> REM Make sure python39 is in the path
> pip install pipenv --user
> REM what goes here to activate the pyenv shell?
> REM what goes here to get ray[default, tune]==1.12.0 and aiohttp==3.8.1?
> python test_script.py
```

**Comment by thomascizeron** [comment]
Same error here (Windows 10, ray==1.12.0, venv), managed to work around by using a global python install. 
For info that was my PATH variable when I encountered this issue:
`PATH=%USERPROFILE%\AppData\Local\Microsoft\WindowsApps;%PyCharm%;C:\Users\<username>\AppData\Local\Programs\Microsoft VS Code\bin;`

**Comment by mattip** [comment]
The fact that it works in a global install but not in "Windows 10, ray==1.12.0, venv" means that in order to fix this I need to know what you mean by that telegraphic phrase. How exactly did you set things up?

**Comment by Jet132** [comment]
I installed Python using Chocolatey: https://community.chocolatey.org/packages/python/3.9.7
I'll try the python.org installation in the next few days.

**Comment by czgdp1807** [comment]
Ah! I see. I use conda environment on my windows to run your code. May be that's why I never received the error which you faced.

You can try using conda environment and let us know if you still face the same error. This might help in finding a point of difference and guide the investigation in a certain direction. Thanks.

**Comment by Jet132** [comment]
Tried a few things with the python.org and Anaconda distributions and found some interesting results.

| Method | Versions | Missing `redis` | Works |
| :-- | :-- | :--: | :--: |  
| User installed Pipenv from global python.org install | <ul><li>pipenv `2022.5.2`</li><li>python `3.9.12`</li></ul> | Yes | No |
| Venv from global python.org install | <ul><li>python `3.9.12`</li></ul> | Yes | No |
| Global python.org install with user installed packages | <ul><li>python `3.9.12`</li></ul> | Yes | Yes |
| Non-base env from global miniconda install | <ul><li>conda `4.11.0`</li><li>python `3.9.12`</li></ul> | Yes | Yes |

By user installed I mean `pip install --user <package>`. While testing I noticed that there also is `C:\Users\<user>\AppData\Local\Microsoft\WindowsApps\python.exe` in PATH, but as it is a later entry it doesn't get used. 

btw. @mattip your script does error if you extend `f`s runtime to 30 seconds using for example `time.sleep(30)`.

**Comment by mattip** [comment]
Thanks @Jet132 this is very helpful. Could you list the commands you used for the first two so that I can try to reproduce?

**Comment by Jet132** [comment]
The first one is documented in the initial issue post. The second one goes as follows:

```shell
python -m venv .venv
.\.venv\Scripts\activate.ps1
pip install ray[default,tune] redis
python .\main.py
```

**Comment by mattip** [comment]
This is really strange. I cannot replicate your failures: neither the failure to see redis (I did not need to add it at all to the install command), nor the failure to run the test program, nor do I get an error when I add a `time.sleep(30)` to `f()`. Do you have any `PYTHON*` environment variables set (Does `set PYTHON` print anything)?

**Comment by Jet132** [comment]
Sorry for the late reply. I was busy last week. No, there are no environment variable names containing `PYTHON`.

I've just tested a clean `3.9.12` python.org default installation (except for enabling the "Add to PATH" option) on a second machine. 
It yielded the same results with `redis` having to be added manually and it erroring out after 30 seconds with `GrpcUnavailable`. Note that I used the following command `pipenv install ray[default] redis` as `ray[tune]` is not needed to reproduce the issue. `sys.path` had the following contents:
```python
['C:\\Users\\<user>\\.virtualenvs\\Test-erJoHnEs\\lib\\site-packages\\ray\\thirdparty_files',
'C:\\Users\\<user>\\Downloads\\Test',
'C:\\Users\\<user>\\Downloads\\Test',
'C:\\Users\\<user>\\.virtualenvs\\Test-erJoHnEs\\lib\\site-packages\\ray\\pickle5_files',
'C:\\Users\\<user>\\.virtualenvs\\Test-erJoHnEs\\lib\\site-packages\\ray\\workers',
'C:\\Users\\<user>\\AppData\\Local\\Programs\\Python\\Python39\\python39.zip',
'C:\\Users\\<user>\\AppData\\Local\\Programs\\Python\\Python39\\DLLs',
'C:\\Users\\<user>\\AppData\\Local\\Programs\\Python\\Python39\\lib',
'C:\\Users\\<user>\\AppData\\Local\\Programs\\Python\\Python39',
'C:\\Users\\<user>\\.virtualenvs\\Test-erJoHnEs',
'C:\\Users\\<user>\\.virtualenvs\\Test-erJoHnEs\\lib\\site-packages']
```

Also I've check out the Microsoft Store installation at `C:\Users\<user>\AppData\Local\Microsoft\WindowsApps\python[3].exe` and it just opens the Python 3.10 store page. Meaning, it isn't actually installed and clearly doesn't get called.

**Comment by mattip** [comment]
The missing redis was reported in #23982, and fixed in #23991. You are correct that `pip install `ray[serve]` alone is broken for the 1.12.0 package.

**Comment by mattip** [comment]
xref #24905, which has a pure-python fix

**Comment by Jet132** [comment]
Will try out the new dev version once I have time and there is a passing build. 
Out of curiosity: Were you able to reproduce the issue?

**Comment by mattip** [comment]
> Were you able to reproduce the issue?

Yes. It seems earlier I was focusing on the wrong areas: the redis thing was a false flag, and the agent issue only shows up after a 30 second timeout.

**Comment by mattip** [comment]
Looking back, you had the `time.sleep()` in your scripts that would have triggered the failure, I must have missed that.

**Comment by mattip** [comment]
@Jet132 if you could check with a [nightly build](https://docs.ray.io/en/master/ray-overview/installation.html#daily-releases-nightlies) that contains the fix from #24968, it would be nice to know for certain that this problem is fixed.

**Comment by Jet132** [comment]
Sorry forgot about it. Also never quite sure if the nightlies only get updated once a commit passed.

Anyway, the redis and timeout problems are both gone. 👍 Thank you very much for the proactive responses and the fix.
